### ISSUE TITLE
Events endStage function does not remove player from Context activePlayers list

### ISSUE THREAD

**Issue by bladx** [boardgameio/boardgame.io#854]
Hello,

I have a move where at the end, if a certain condition is fulfilled I call `ctx.events.endStage();`.
But for some reasons, when I log the `ctx.activePlayers` after that particular line, I still see my player in the list.

My objectif behind this was to have a game where a turn is split into 2 phases where all players can do actions.
This is a bit like 7 wonders, like:
-  there is not really an **current** player
- all players play simultaneously (choosing a card) - Stage 1
- then only when all players chose, they decide to either build it or sell it or use it to build wonder (again simultaneously) - Stage 2
- then the turn end and we restart from Stage 1
The difference is that I want to display players choice between Stage A and Stage B. This means I want to specifically know when Stage A is over to display all players info before entering in stage B.

If calling the `ctx.events.endStage();` actually removed the player from `ctx.activePlayers` then I should be able to call the function I want at the end of stage 1 and/or Stage 2.

I found out that when I set a moveLimit then before entering to onMove, my customer is deleted from `ctx.activePlayers` (if he reached the move limit of course). To me, we should have the same behaviour calling `ctx.events.endStage();`.
I precise that my stages do not have a given move limit.

**Comment by delucis** [comment]
Hi @bladx! Where are you logging `ctx.activePlayers`? It’s important to note that the player will only be removed from the active players *after* your move function completes. See [the docs](https://boardgame.io/documentation/#/events?id=triggering-an-event-from-game-logic):

> Events are queued up and triggered after a move. Any changes you make to G will be applied before events are triggered, even if the event is called first in your move function.

It’s not currently as straightforward as it should be to write logic like you describe. You might want to check out #769 for an example of how to do this kind of thing. #608 is also requesting something that could help with these scenarios, but no-one has implemented that yet.

**Comment by bladx** [comment]
Hi Chris,

Thank you very much for the quick answer!
I was logging `ctx.activePlayers` in the first line of my onMove function.
With your docs quote that explains why I had this behaviour.
Sorry for that, I should have missed this part from the docs.

Anyway, I will create a function that kind of override the `ctx.events.endStage()` method as you suggested to Alvaro in #769.
This is more a workaround than a solution to me, but that should allow me to achieve what I want.

A maybe nice feature would be to have a onEvent listener that would work as onMove but would be triggered after a move.
I don't know if that would be the best solution but I feel like a trigger is missing to know about events changes.
What do you think about this?

**Comment by delucis** [comment]
@bladx Oh, it’s definitely a workaround, for sure 😄 

> A maybe nice feature would be to have a onEvent listener

We already have hooks for when turns and phases begin and end, what is missing is the same for stages as requested in #608. That could help because you could have something like:

```js
stages: {
  foo: {
    onEnd: (G, ctx) => {
      if (ctx.activePlayers === null) ctx.endTurn();
    },
  },
},
```

Also `turn.endIf` should be updated to run after active players has been updated as requested in #769.

One more thing I forgot that might help in your case is using [the `next` option when setting active players](https://boardgame.io/documentation/#/stages?id=advanced):

```js
turn: {
  activePlayers: {
    all: 'stageOne',
    next: {
      all: 'stageTwo',
    },
  },
},
```

In this case, all players are set to `stageOne`, then once `ctx.activePlayers` becomes empty (using `endStage`), the `next` option will be used, setting all players to `stageTwo`.

**Comment by bladx** [comment]
Hi again,

Yes, I did try that but unfortunately it does not fit my needs as I want to perform a function at the end of each stage, plus end the turn at the end of stageTwo. But thank you anyway.

Indeed both requests #608 and #769 would help me a lot. Do you know if this is plan in a soon version?
I will use the hack you gave me earlier until there is a better solution.

**Comment by delucis** [comment]
> Do you know if this is plan in a soon version?

Depends if someone has time to work on it, so I can’t guarantee when, but hopefully it won’t be forever!

I’ll close this for now as I think the requested features/fixes are covered by #608 and #769, but feel free to ask if you have any further questions.
